### ISSUE TITLE
Memory leak with dragDropManager as retainer object

### ISSUE THREAD

**Issue by canastro** [react-dnd/react-dnd#1532]
**Describe the bug**

Possible memory leak with react-dnd. 

In a app with react-router-dom when I navigate away of a screen with react-dnd I still see a lot of my models instances, unrelated components (not child neither sibilings of my component containing the DnDProvider) and `detached dom elements` with references to `dragDropManager`. 

At the current state the memory usage after I leave the screen with the DnD is around 49MB, if I comment out the DND components (which is now as simple as possible: provider + 1 drop zone) it goes to 28MB.

I was able to recreate a small sample which only shows the leak of detached dom elements with the references to `dragDropManager`.

I couldn't yet replicate the scenario where multiple components are kept in memory with dragDropManager as retainer object. But here is a picture of the retainer in one of my components:
<img width="624" alt="Screenshot 2019-09-11 at 10 37 34" src="https://user-images.githubusercontent.com/1267623/64686279-55173f00-d480-11e9-802a-80d725ab6898.png">

**Reproduction**

**Video showing the possible memory leak:** https://monosnap.com/file/1aKyq523ZkKY1gO39QS05yjBmzEBB0

1. Clone https://github.com/canastro/dnd-sample
3. Run sample in production mode with `npm run prod`
4. Open http://localhost:5000
5. Go to the DND page and navigate back and sometimes between the other two pages
6. Take heap snapshot, check for dettached div elements, and check a retainer related with react-dnd

**Expected behavior**
All references to react-dnd should be cleaned up properly.

**Desktop (please complete the following information):**

- OS: macOS Movaje
- Browser: chrome
- Version:  76.0.3809.132

**Comment by canastro** [comment]
@gaearon since you're familiar with both react an react-dnd can you point out if this issue is something related with DND or if it fits the points already identified in the [memory leaks umbrella issue of react](https://github.com/facebook/react/issues/16087)?

**Comment by canastro** [comment]
I've found two different ways to address this, one in react-dnd's source code (but probably this would mess up 1001 different use cases) and another in user-land.

In react-dnd source code I would add something roughly like this in the DndProvider:

```js
React.useEffect(() => {
	return () => {
		if ('manager' in props) {
			return;
		}

		const context = props.context || getGlobalContext();
		context[instanceSymbol] = null;
	}
}, []);
```

In user land I would have to use the DND in one of two following ways:

```jsx
const manager = useRef(createDndContext(HTML5Backend));

<DndProvider manager={manager.current.dragDropManager}>
...
</DndProvider>
```


@darthtrevino I'm going with user-land approach, do you think this might be a wrong usage of react-dnd? Is this something that should be fixed (if its considered a issue) on react-dnd side?

**Comment by darthtrevino** [comment]
I see, so detaching the DndProvider element would wipe it from he global context - that sounds reasonable to me

**Comment by canastro** [comment]
@darthtrevino I can try, but I'm not sure of the issues related with the usage of multiple providers. I opened up this PR: https://github.com/react-dnd/react-dnd/pull/1543 check it out if this is something like you had in mind

**Comment by canastro** [comment]
I can confirm that this issue has been fixed with version 9.4.0:

Before:
<img width="183" alt="heap snapshots from version prior to 9.4.0 show that no memory from previous route was released" src="https://user-images.githubusercontent.com/1267623/65406649-16ff1100-ddd8-11e9-863d-bc1f2dc683e1.png">

After:
<img width="167" alt="heap snapshots from version 9.4.0 show that all memory was released" src="https://user-images.githubusercontent.com/1267623/65406653-19fa0180-ddd8-11e9-9021-937a39847343.png">
